fix(core): Allow passing AbstractType to the inject function

mitchellwills opened the pull request:
This is a type only change that replaces `Type<T>|InjectionToken<T>` with
`Type<T>|AbstractType<T>|InjectionToken<T>` in the injector.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

`inject` instruction does not accept `AbstractType` as a token.

Issue Number: N/A


## What is the new behavior?

`inject` instruction accepts `AbstractType` as a token. This is a type only change (should have no runtime impact)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Potentially `Type<T>|AbstractType<T>|InjectionToken<T>` should become a type alias, but i'm not sure what it should be called.

mitchellwills committed:
fix(core): Allow passing AbstractType to the inject function

This is a type only change that replaces `Type<T>|InjectionToken<T>` with
`Type<T>|AbstractType<T>|InjectionToken<T>` in the injector.

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

petebacondarwin reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
Hi @mitchellwills, thanks for creating this PR 👍 

I did a quick search in Angular codebase and found a couple more places where the `AbstractType` is not included (both in the current state of the codebase and in your PR):
- https://github.com/angular/angular/blob/master/packages/core/src/di/injector.ts#L157
- https://github.com/angular/angular/blob/master/packages/core/src/di/r3_injector.ts#L183
- https://github.com/angular/angular/blob/master/packages/core/src/di/r3_injector.ts#L406
- https://github.com/angular/angular/blob/master/packages/core/src/render3/component_ref.ts#L85
- https://github.com/angular/angular/blob/master/packages/core/src/render3/interfaces/injector.ts#L137
- https://github.com/angular/angular/blob/master/packages/core/src/render3/interfaces/injector.ts#L241
- https://github.com/angular/angular/blob/master/packages/examples/core/di/ts/injector_spec.ts#L15

Could you please have a quick look to see if these places also require type change?

> Potentially Type<T>|InjectionToken<T>|AbstractType<T> should become a type alias, but i'm not sure what it should be called.

I really like this idea, could you please create a new ticket (feature request) to capture this proposal?

Thank you.

AndrewKushnir reviewed (commented):
Hi @mitchellwills, thanks for creating this PR 👍 

I did a quick search in Angular codebase and found a couple more places where the `AbstractType` is not included (both in the current state of the codebase and in your PR):
- https://github.com/angular/angular/blob/master/packages/core/src/di/injector.ts#L157
- https://github.com/angular/angular/blob/master/packages/core/src/di/r3_injector.ts#L183
- https://github.com/angular/angular/blob/master/packages/core/src/di/r3_injector.ts#L406
- https://github.com/angular/angular/blob/master/packages/core/src/render3/component_ref.ts#L85
- https://github.com/angular/angular/blob/master/packages/core/src/render3/interfaces/injector.ts#L137
- https://github.com/angular/angular/blob/master/packages/core/src/render3/interfaces/injector.ts#L241
- https://github.com/angular/angular/blob/master/packages/examples/core/di/ts/injector_spec.ts#L15

Could you please have a quick look to see if these places also require type change?

> Potentially Type<T>|InjectionToken<T>|AbstractType<T> should become a type alias, but i'm not sure what it should be called.

I really like this idea, could you please create a new ticket (feature request) to capture this proposal?

Thank you.

AndrewKushnir commented:
@googlebot ping.

atscott reviewed (approved):
Reviewed-for: public-api

alxhub commented on packages/core/src/di/injector_compatibility.ts:
```diff
@@ -110,9 +113,11 @@ export function injectInjectorOnly<T>(
-export function ɵɵinject<T>(token: Type<T>|InjectionToken<T>): T;
-export function ɵɵinject<T>(token: Type<T>|InjectionToken<T>, flags?: InjectFlags): T|null;
-export function ɵɵinject<T>(token: Type<T>|InjectionToken<T>, flags = InjectFlags.Default): T|null {
+export function ɵɵinject<T>(token: Type<T>|InjectionToken<T>|AbstractType<T>): T;
```
It's just a nitpick, but I would suggest reordering `AbstractType` and `InjectionToken`:
```suggestion
export function ɵɵinject<T>(token: Type<T>|AbstractType<T>|InjectionToken<T>): T;
```
as they're two sides of the same coin, vs `InjectionToken` which is more of a foreign currency ;)

alxhub reviewed (approved):
Reviewed-for: public-api

LGTM with one suggestion

atscott reviewed (approved):
Adding approval for fw-core

petebacondarwin commented:
@mitchellwills - are you still interested in working on this PR? If so, could you rebase and deal with the comments? Thanks!

mitchellwills commented:
yah, sorry for the delay (I worked around it so this dropped in priority for me). I'll rebase and address the comments next week

AndrewKushnir commented:
FYI, related ticket: #23611.

mitchellwills commented:
I think i've addressed the comments from above: rebase, additional replacements, reorder to `AbstractType<T>|InjectionToken<T>`. 

I also filed https://github.com/angular/angular/issues/39792

AndrewKushnir commented:
[Presubmit](http://test/OCL:343551359:BASE:343551408:1605906780367:ac083ed9).

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:343551359:BASE:343980727:1606192448865:7028c338).

AndrewKushnir commented:
FYI, global presubmit indicated no problems related to this change, so I'm adding it to the merge queue. Note: I've changed the target from "patch" to "minor" though since this PR can also be considered as a feature (that more types are now allowed in the inject function).
